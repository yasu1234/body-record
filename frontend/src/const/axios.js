import axios from "axios";
import Cookies from "js-cookie";

const baseURL = import.meta.env.VITE_APP_API_BASE;

const axiosInstance = axios.create({
  baseURL,
  headers: {
    "Content-Type": "application/json",
  },
});

axiosInstance.interceptors.request.use((config) => {
  config.headers["access-token"] = Cookies.get("accessToken");
  config.headers["client"] = Cookies.get("client");
  config.headers["uid"] = Cookies.get("uid");

  return config;
});

export default axiosInstance;
