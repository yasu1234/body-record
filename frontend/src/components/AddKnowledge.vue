<script setup>
import { ref, onMounted } from 'vue';
import axios from 'axios';
import Cookies from 'js-cookie';

import DropFile from './DropFile.vue'
import Header from './Header.vue'

const title = ref("");
const knowledge = ref("");
const files = ref([]);

function dateChange(event) {
    recordDate.value = event
}

function onFileChange(event) {
    files.value = [...event];
}

const registerRecord = async () => {
  try {
        const formData = new FormData();
        formData.append('title', title.value);
        formData.append('content', knowledge.value);

        for (const file of files.value) {
            formData.append('images', file);
        }

        const res = await axios.post(import.meta.env.VITE_APP_API_BASE + '/api/v1/knowledges', formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        console.log({ res })
    } catch (error) {
        console.log({ error })
    }
}
</script>

<template>
    <Header />
    <div class="editor">
        <label class="itemLabel">タイトル</label>
        <input id="title" type="text" v-model="title">
        <textarea name="knowledge" rows="20" v-model="knowledge"></textarea>
    </div>
    <div class="relationImages">
        <p class="inputTitle">関連画像</p>
        <DropFile @change="onFileChange"/>
    </div>
    <div class="relationImages">
        <button class="registerButton" @click="registerRecord">登録する</button>
    </div>
</template>

<style>
.editor{
   padding: 30px;
 }

 input[type=text] {
  width: 100%;
  padding: 12px 12px;
  margin: 8px 0;
  box-sizing: border-box;
  border: 1px solid #ccc;
  border-radius: 4px;
}
.time-list {
  display: flex;
  justify-content: space-between;
  padding-left: 30px;
  padding-right: 30px;
}
.time-list .item {
  width: 50%;
  margin: 0;
  padding: 10px;
  box-sizing: border-box;
}
.time-list .item .inputTitle {
  margin: 5px 0 0;
  padding: 0;
  font-size: 16px;
}
.relationImages {
    padding: 20px;
}
.registerButton{
    background: #ffa500;
    color: white;
    font-size:16px;
    font-weight:bold;
}
</style>