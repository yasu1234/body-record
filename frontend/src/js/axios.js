import axios from "axios";
import Cookies from "js-cookie";

const baseURL = import.meta.env.VITE_APP_API_BASE;

const axiosInstance = axios.create({
  baseURL,
  headers: {
    "Content-Type": "application/json",
  },
});

axiosInstance.interceptors.request.use((config) => {
  config.headers["access-token"] = Cookies.get("accessToken");
  config.headers["client"] = Cookies.get("client");
  config.headers["uid"] = Cookies.get("uid");

  return config;
});

const setupInterceptors = (router) => {
  axiosInstance.interceptors.response.use(
    response => response,
    error => {
      // 共通のエラー処理
      if (error.response.status === 404) {
        router.push({ name: 'NotFound' })
      }

      // エラーをキャッチしてreturnする
      return Promise.reject(error)
    }
  )
}

export { axiosInstance, setupInterceptors }
