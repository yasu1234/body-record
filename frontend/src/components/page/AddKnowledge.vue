<script setup>
import { ref } from "vue";
import { useRouter } from "vue-router";
import Toast from "primevue/toast";
import { useToast } from "primevue/usetoast";
import { toastService } from "../../const/toast.js";
import FloatLabel from "primevue/floatlabel";
import InputText from "primevue/inputtext";
import Textarea from "primevue/textarea";
import axiosInstance from "../../const/axios.js";

import DropFile from "../atom/DropFile.vue";
import Header from "../layout/Header.vue";
import TabMenu from "../layout/TabMenu.vue";

const router = useRouter();
const toast = useToast();
const toastNotifications = new toastService(toast);

const title = ref("");
const knowledge = ref("");
const files = ref([]);

function onFileChange(event) {
  files.value = [...event];
}

const registerKnowledge = async () => {
  try {
    const formData = new FormData();
    formData.append("knowledge[title]", title.value);
    formData.append("knowledge[content]", knowledge.value);

    for (const file of files.value) {
      formData.append("knowledge[images]", file);
    }

    const res = await axiosInstance.post("/api/v1/knowledges", formData);
    toastNotifications.displayInfo("登録しました", "");
    setTimeout(async () => {
      showKnowledgeDetail(res.data.knowledge);
    }, 3000);
  } catch (error) {
    let errorMessages = "";
    if (error.response.status === 422) {
      if (Array.isArray(error.response.data.errors)) {
        errorMessages += error.response.data.errors.join("\n");
      }
    }
    toastNotifications.displayError(
      "ノウハウの追加に失敗しました",
      errorMessages
    );
  }
};

const showKnowledgeDetail = (item) => {
  router.push({ name: "KnowledgeDetail", params: { id: item.id } });
};
</script>

<template>
  <Header />
  <TabMenu />
  <Toast position="top-center" />
  <div class="p-7">
    <FloatLabel class="mt-5">
      <InputText v-model="title" class="input-width" />
      <label>タイトル</label>
    </FloatLabel>
    <FloatLabel class="mt-5">
      <Textarea v-model="knowledge" rows="20" class="input-width" />
      <label>詳細</label>
    </FloatLabel>
  </div>
  <div class="p-5">
    <p class="inputTitle">関連画像(5枚まで登録できます)</p>
    <div v-for="i in 5">
      <DropFile @change="onFileChange" :index="i" class="mt-3" />
    </div>
  </div>
  <div class="p-5 text-center">
    <button class="add-knowledge-button" @click="registerKnowledge">
      登録する
    </button>
  </div>
</template>

<style scoped>
.input-width {
  width: 100%;
  padding: 10px;
}
.add-knowledge-button {
  font-size: 16px;
  font-weight: bold;
}
</style>
