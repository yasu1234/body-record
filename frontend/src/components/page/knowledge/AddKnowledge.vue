<template>
  <Header />
  <TabMenu />
  <Toast position="top-center" />
  <form class="p-7" @submit.prevent="registerKnowledge">
    <FloatLabel class="mt-5">
      <InputText v-model="title" class="input-width" />
      <label>タイトル</label>
    </FloatLabel>
    <div class="mt=2.5">
      <KnowledgeContentInput
        :knowledgeContent="knowledge"
        @content-edit="contentEdit"
      />
    </div>
    <div class="markdown-info">
      マークダウンを一部使用できます。詳しくは<a
        href="https://kumaapp.amebaownd.com/posts/54283163"
        target="_blank"
        rel="noopener noreferrer"
        class="info-link"
        >こちら</a
      >をご覧ください。<br />
      特に複数行の改行は使うことが多いと思うので必ずご確認ください
    </div>
    <div class="p-5">
      <h2>関連画像(5枚まで登録できます)</h2>
      <div class="file-input-container">
        <div v-for="i in 5">
          <DropFile @change="onFileChange" :index="i" class="mt-3" />
        </div>
      </div>
    </div>
    <div class="p-5 text-center">
      <button class="add-knowledge-button">登録する</button>
    </div>
  </form>
</template>

<script setup>
import { ref } from "vue";
import { useRouter } from "vue-router";
import Toast from "primevue/toast";
import { useToast } from "primevue/usetoast";
import { toastService } from "../../../js/toast.js";
import FloatLabel from "primevue/floatlabel";
import InputText from "primevue/inputtext";
import { axiosInstance } from "../../../js/axios.js";

import DropFile from "../../atom/DropFile.vue";
import Header from "../../layout/Header.vue";
import TabMenu from "../../layout/TabMenu.vue";
import KnowledgeContentInput from "../../layout/KnowledgeContentInput.vue";

const router = useRouter();
const toast = useToast();
const toastNotifications = new toastService(toast);

const title = ref("");
const knowledge = ref("");
const files = ref([...Array(5)]);

const registerKnowledge = async () => {
  try {
    const formData = new FormData();
    formData.append("knowledge[title]", title.value);
    formData.append("knowledge[content]", knowledge.value);

    for (const file of files.value) {
      if (file != null) {
        formData.append("knowledge[images][]", file);
      }
    }

    const res = await axiosInstance.post("/api/v1/knowledges", formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    });
    toastNotifications.displayInfo("登録しました", "");
    setTimeout(async () => {
      showKnowledgeDetail(res.data.knowledge);
    }, 3000);
  } catch (error) {
    if (error.response == null) {
      toastNotifications.displayError("記事の編集に失敗しました", "");
      return;
    }

    let errorMessage = "";

    if (error.response.status === 422) {
      if (Array.isArray(error.response.data.errors)) {
        errorMessage += error.response.data.errors.join("\n");
      } else {
        errorMessage = error.response.data.errors;
      }
    } else if (error.response.status === 401) {
      errorMessage = "ログインしてください";
    }

    toastNotifications.displayError("記事の追加に失敗しました", errorMessage);
  }
};

const onFileChange = (event, index) => {
  files.value[index - 1] = event;
};

const contentEdit = (editContent) => {
  knowledge.value = editContent;
};

const showKnowledgeDetail = (item) => {
  router.push({ name: "KnowledgeDetail", params: { id: item.id } });
};
</script>

<style scoped>
.input-width {
  width: 100%;
  padding: 10px;
}
.add-knowledge-button {
  font-size: 16px;
  font-weight: bold;
  padding: 10px 50px;
}
.markdown-info {
  font-size: 14px;
  color: #333;
}
.info-link {
  color: #ffa500;
  text-decoration: underline;
  cursor: pointer;
}
.info-link:hover {
  color: #ffa500;
}
</style>
