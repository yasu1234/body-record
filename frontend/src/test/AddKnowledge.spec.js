import { shallowMount } from '@vue/test-utils'
import axios from 'axios'
import AddKnowledge from '@/components/page/AddKnowledge.vue'

jest.mock('axios')
jest.mock('vue-router', () => {
    const router = {
      push: jest.fn(),
      pathname: '/',
      route: '',
      asPath: '',
      query: {},
      events: {
        on: jest.fn(),
        off: jest.fn(),
        emit: jest.fn(),
      },
    };
    return {
      useRouter: jest.fn().mockReturnValue(router),
      useRoute: jest.fn().mockReturnValue({
        params: {
          id: 1
        }
      })
    };
  });

  jest.mock('primevue/usetoast', () => ({
    useToast: jest.fn(() => ({
      add: jest.fn(),
      removeAll: jest.fn()
    }))
  }))

describe('AddKnowledge', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(AddKnowledge, {
        global: {
          stubs: ["router-link", "router-view"]
        }
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('SnapShotテスト初期表示', () => {
    expect(wrapper.element).toMatchSnapshot();
  })
})
