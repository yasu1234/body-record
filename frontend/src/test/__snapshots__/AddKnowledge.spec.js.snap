// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddKnowledge SnapShotテスト初期表示 1`] = `
<div
  data-v-app=""
>
  
  <header-stub />
  <tab-menu-stub />
  <toast-stub
    position="top-center"
  />
  <div
    class="p-7"
  >
    <float-label-stub
      class="mt-5"
    />
    <float-label-stub
      class="mt-5"
    />
  </div>
  <div
    class="p-5"
  >
    <h2>
      関連画像(5枚まで登録できます)
    </h2>
    <div
      class="file-input-container"
    >
      
      <div>
        <drop-file-stub
          class="mt-3"
          index="1"
        />
      </div>
      <div>
        <drop-file-stub
          class="mt-3"
          index="2"
        />
      </div>
      <div>
        <drop-file-stub
          class="mt-3"
          index="3"
        />
      </div>
      <div>
        <drop-file-stub
          class="mt-3"
          index="4"
        />
      </div>
      <div>
        <drop-file-stub
          class="mt-3"
          index="5"
        />
      </div>
      
    </div>
  </div>
  <div
    class="p-5 text-center"
  >
    <button
      class="add-knowledge-button"
    >
       登録する 
    </button>
  </div>
  
</div>
`;
